js/components/authoring.js

```diff
@@ -43,36 +47,38 @@ export default class Authoring extends PureComponent {
     const { step } = this.state
     const { setOption } = this.props
     if (step === 2) {
-      setOption('interaction', 'force')
-      setOption('selectableInteractions', [ 'force', 'none' ])
-      this.setState({step: 3})
+      this.snapshotStep()
+      this.setStep3()
     } else if (step === 3) {
-      setOption('interaction', 'none')
-      setOption('selectableInteractions', [])
-      setOption('colormap', 'plate')
-      this.setState({step: 4})
+      if (config.densityStepEnabled) {
+        this.snapshotStep()
+        setOption('interaction', 'none')
+        setOption('selectableInteractions', [])
+        setOption('colormap', 'plate')
+        this.setState({step: 4})
+      } else {
+        this.endAuthoring()
+      }
     } else if (step === 4) {
       const { setDensities } = this.props
       setDensities(this.state.intermediateDensities)
       setOption('colormap', 'topo')
 
-      setOption('authoring', false)
-      setOption('playing', true)
-      setOption('interaction', 'none')
-      setOption('colormap', config.colormap)
-      setOption('renderBoundaries', config.renderBoundaries)
-      setOption('renderForces', config.renderForces)
-      setOption('selectableInteractions', config.selectableInteractions)
-      this.setState({step: 5})
+      this.endAuthoring()
     }
   }
 
   handleBackButtonClick () {
     const { step } = this.state
+    const { restoreLabeledSnapshot } = this.props
     if (step === 2) {
       this.unloadModel()
     } else if (step === 3) {
+      restoreLabeledSnapshot(STEP_NAMES[2])
       this.setStep2()
+    } else if (step === 4) {
+      restoreLabeledSnapshot(STEP_NAMES[3])
```

Comment:
I opted for restoring steps by number explicitly, rather than making a function that restores the "last" step. In the future, if multiple steps can be skipped, going to the "last" numerical step might be incorrect.